[ONNX] Refactor AvgPool to support dynamic shapes

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105683

In #87892, to pick up the corner cases found in #71549, the PR falls back the implementation of AvgPool to the way opset 9 implementing. However, it introduces a regression on dynamic shape cases found in #101397. This PR refactors the AvgPool op with the same implementation we have in onnxscript: https://github.com/microsoft/onnxscript/pull/754.

However, the corner case with `count_include_pad` remains unsolved in onnxruntime: https://github.com/microsoft/onnxruntime/issues/16203. The calculuation on the last value of each dimension is different between ORT and PyTorch. But the fix can be proved in: https://github.com/microsoft/onnxruntime/pull/16752, and it supports AvgPool since opset19.

titaiwangms committed:
[ONNX] Refactor AvgPool to support dynamic shapes

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1490,29 +1492,31 @@ def test_avgpool_1d_ceil(self):
         x = torch.randn(20, 16, 50, 32)
         self.run_test(model, x)
 
+    # NOTE: ceil_mode is not included in the test, because of
```
```suggestion
    # TODO: ceil_mode is not included in the test, because of
```

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1490,29 +1492,31 @@ def test_avgpool_1d_ceil(self):
         x = torch.randn(1, 1, 7)
         self.run_test(model, x)
 
+    # NOTE: ceil_mode is not included in the test, because of
```
```suggestion
    # TODO: ceil_mode is not included in the test, because of
```

thiagocrepaldi reviewed (approved):
LGTM.

IIUC, the `ceil_mode` failure is a bug on ORT side, which will be addressed there?

titaiwangms committed:
Update on "[ONNX] Refactor AvgPool to support dynamic shapes"


In #87892, to pick up the corner cases found in #71549, the PR falls back the implementation of AvgPool to the way opset 9 implementing. However, it introduces a regression on dynamic shape cases found in #101397. This PR refactors the AvgPool op with the same implementation we have in onnxscript: https://github.com/microsoft/onnxscript/pull/754.

However, the corner case with `ceil_mode` remains unsolved in onnxruntime: https://github.com/microsoft/onnxruntime/issues/16203. The calculuation on the last value of each dimension is different between ORT and PyTorch. But the fix can be proved in: https://github.com/microsoft/onnxruntime/pull/16752, and it supports AvgPool since opset19.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor AvgPool to support dynamic shapes"


In #87892, to pick up the corner cases found in #71549, the PR falls back the implementation of AvgPool to the way opset 9 implementing. However, it introduces a regression on dynamic shape cases found in #101397. This PR refactors the AvgPool op with the same implementation we have in onnxscript: https://github.com/microsoft/onnxscript/pull/754.

However, the corner case with `ceil_mode` remains unsolved in onnxruntime: https://github.com/microsoft/onnxruntime/issues/16203. The calculuation on the last value of each dimension is different between ORT and PyTorch. But the fix can be proved in: https://github.com/microsoft/onnxruntime/pull/16752, and it supports AvgPool since opset19.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor AvgPool to support dynamic shapes"


In #87892, to pick up the corner cases found in #71549, the PR falls back the implementation of AvgPool to the way opset 9 implementing. However, it introduces a regression on dynamic shape cases found in #101397. This PR refactors the AvgPool op with the same implementation we have in onnxscript: https://github.com/microsoft/onnxscript/pull/754.

However, the corner case with `ceil_mode` remains unsolved in onnxruntime: https://github.com/microsoft/onnxruntime/issues/16203. The calculuation on the last value of each dimension is different between ORT and PyTorch. But the fix can be proved in: https://github.com/microsoft/onnxruntime/pull/16752, and it supports AvgPool since opset19.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor AvgPool to support dynamic shapes"


In #87892, to pick up the corner cases found in #71549, the PR falls back the implementation of AvgPool to the way opset 9 implementing. However, it introduces a regression on dynamic shape cases found in #101397. This PR refactors the AvgPool op with the same implementation we have in onnxscript: https://github.com/microsoft/onnxscript/pull/754.

However, the corner case with `ceil_mode` remains unsolved in onnxruntime: https://github.com/microsoft/onnxruntime/issues/16203. The calculuation on the last value of each dimension is different between ORT and PyTorch. But the fix can be proved in: https://github.com/microsoft/onnxruntime/pull/16752, and it supports AvgPool since opset19.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
